RIPS/rip-9998.md

```diff
@@ -0,0 +1,306 @@
+
+```
+
+The `transactions` parameter is interpreted as an array of `TransactionType4` structs.
+
+This call returns a value used as `aggregatedSignature` elsewhere. Aggregator MUST revert this call if it is unable to
+calculate this value.
+
+Note that the mempool MUST only include aggregated AA transactions for `aggregators` that have an implementation
+of `aggregateSignatures` function.
+
+A block builder MAY calculate this value using any alternative way.
+If the aggregator does not implement `aggregateSignatures` function it is up to block builder to add
+support for this `aggregator` and its signature scheme.
+
+### Execution layer validity of a block with aggregated transactions included
+
+The `validateAccountAbstractionTransaction` function's logic is not modified compared to RIP-7560.
+
+On the execution layer, the block transactions validity conditions are extended as follows:
+
+```
+
+if type(tx) is AccountAbstractionTransaction {
+    if (tx.version == AA_AGGREGATED_SIG_TX_VERSION){
+            aggregatedTxs := collectTransactions(tx.transactionCount)
+            aggregatedSignature := aggregateSig(aggregatedTxs)
+            aggregatorInput := ABI.encodeWithSelector('validateSignatures()', aggregatedTxs, aggregatedSignature)
+            gasUsedByAggregator := evm.Call(
+                from: AA_ENTRY_POINT,
+                to: tx.aggreagor,
+                input: aggregatorInput,
+                gas: config.aa_aggregated_sig_validation_gas_limit)
+        else if (tx.version == AA_AGGREGATED_TX_VERSION){
+            // NOTE: we subtract the share of aggregated tx signature valdiation gas used from the validation gas limit
+            tx.validationGasLimit = tx.validationGasLimit - (gasUsedByAggregator / tx.transactionCount)
+            validateAccountAbstractionTransaction(tx)
+        }
+    }
+}
+
+// note: make sure all transactions are accounted for
+assert all_type_x_version_1_transactions_validated()
+
+// note: 'aggregated signature data' transactions do not pay their own gas
+// make sure no such transaction is left 'hanging' in a block without paying 'child' transactions
+assert all_type_x_version_1_transactions_gas_paid_for()
+
+```
+
+### Passing cost savings to users via gas cost adjustments
+
+The RIP-7560 charges the signature validation cost as part of a transaction gas usage.
+It also treats transaction `signature` parameter as an arbitrary length byte arrays input. Its cost is calculated in the
+same way the transaction `data` cost is calculated for previous transaction types.
+
+With signature aggregation, however, the individual transactions no longer carry their own signature. They also share
+the cost of validating the aggregated signature.
+
+Therefore, the gas cost of the aggregated validation transaction is evenly divided by the number of transactions it
+validates and is added to each aggregated transaction's gas cost.
```

Comment:
This incentivizes the builder to make aggregated batches small (e.g. batches of 1 transaction).  It's a zero-sum game - the more the user pays, the more the builder earns.  And in the current model the user has no say in this.

The alternative is that the builder (or whoever sends the AA_TX_TYPE_AGGREGATED_SIG transaction) pays for aggregation directly, and gets compensated by the user transaction, e.g. via `builderFee` (if it's the builder) or via the transaction itself (if it's something like an intent solver).  But even then, there's still information asymmetry that works against the user.  The builder knows how many pending transactions it has for an aggregator, but the user doesn't.  If the user sets `builderFee` too low for the current number of pending transactions, the transaction gets delayed until it is worthwhile to include it.  If the user sets it too high, the savings on big aggregated batches go to the builder rather than the user.

We still need to figure out the incentives model for aggregation in a way that doesn't bring back pre-1559 inefficiencies.